pkg/handler/enqueue_mapped.go

```diff
@@ -79,10 +82,23 @@ func (e *enqueueRequestsFromMapFunc) Generic(ctx context.Context, evt event.Gene
 
 func (e *enqueueRequestsFromMapFunc) mapAndEnqueue(ctx context.Context, q workqueue.RateLimitingInterface, object client.Object, reqs map[reconcile.Request]empty) {
 	for _, req := range e.toRequests(ctx, object) {
-		_, ok := reqs[req]
-		if !ok {
-			q.Add(req)
-			reqs[req] = empty{}
+		if _, ok := reqs[req]; ok {
+			// reqs is a map of requests to avoid enqueueing the same request multiple times.
+			continue
+		}
+		// If the request doesn't specify a cluster, use the cluster from the context.
+		if req.Cluster == "" && e.cluster != nil {
+			req.Cluster = e.cluster.Name()
 		}
+		// Enqueue the request and track it.
+		q.Add(req)
+		reqs[req] = empty{}
+	}
+}
+
+func (e *enqueueRequestsFromMapFunc) DeepCopyFor(c cluster.Cluster) EventHandler {
+	return &enqueueRequestsFromMapFunc{
+		cluster:    c,
```

Comment:
Wondering if we need to store the entire Cluster given that we only use the name.

Reply:
Related: at least in `EnqueueRequestForObject` we could additionally store the ClusterName in an additional field to avoid the lookup for every event

(Not sure if the compiler optimizes it away, but still simplifies the implementation)